actions/rules/action_state_to_metastate.go

```diff
@@ -0,0 +1,416 @@
+		}
+		// extract the column's metastate config.
+		if metaState, ok := config[ActionKeyStateToMetastateConfigMetastate]; ok {
+			if metaState == wi.Fields[workitem.SystemMetaState] {
+				// the WIs metastate is already the same as the columns
+				// metastate, so nothing to do.
+				return newContext, *actionChanges, nil
+			}
+			// the metatstate changes, so set it on the WI.
+			changes = act.addOrUpdateChange(actionChanges, workitem.SystemMetaState, wi.Fields[workitem.SystemMetaState], metaState)
+			wi.Fields[workitem.SystemMetaState] = metaState
+			wiDirty = true
+			// next, check if the state needs to change as well from the metastate.
+			if wi.Fields[workitem.SystemState] != mapping[metaState] {
+				// yes, the state changes as well.
+				changes = act.addOrUpdateChange(actionChanges, workitem.SystemState, wi.Fields[workitem.SystemState], mapping[metaState])
+				wi.Fields[workitem.SystemState] = mapping[metaState]
+			}
+		}
+	}
+	// finally, store the new work item state if something changed.
+	if wiDirty {
+		act.storeWorkItem(&wi)
+	}
+	// return to sender
+	return newContext, changes, nil
+}
+
+// OnStateChange is executed when the state changes. It eventually updates the metastate and the boardcolumns.
+func (act ActionStateToMetaState) OnStateChange(newContext convert.ChangeDetector, contextChanges []convert.Change, configuration string, actionChanges *[]convert.Change) (convert.ChangeDetector, []convert.Change, error) {
+	wi, ok := newContext.(workitem.WorkItem)
+	if !ok {
+		return nil, nil, errors.New("Given context is not a WorkItem instance")
+	}
+	// get the mapping.
+	mapping, err := act.getStateToMetastateMap(wi.Type)
+	if err != nil {
+		return nil, nil, err
+	}
+	// create a dirty flag. Note that we can not use len(actionChanges) as this
+	// may contain previous changes from the action chain.
+	wiDirty := false
+	// update the workitem accordingly.
+	if wi.Fields[workitem.SystemMetaState] == mapping[workitem.SystemState] {
+		// metastate remains stable, nothing to do.
+		return newContext, *actionChanges, nil
+	}
+	// otherwise, update the metastate from the state.
+	changes := act.addOrUpdateChange(actionChanges, workitem.SystemMetaState, wi.Fields[workitem.SystemMetaState], mapping[workitem.SystemState])
+	wi.Fields[workitem.SystemMetaState] = mapping[workitem.SystemState]
+	wiDirty = true
+	// next, get the columns of the workitem and see if these needs to be updated.
+	boards, err := act.loadWorkItemBoardsBySpaceID(wi.SpaceID)
+	if err != nil {
+		return nil, nil, err
+	}
+	// next, check which boards are relevant for this WI.
+	groups, err := act.loadWorkItemTypeGroupsBySpaceID(wi.SpaceID)
+	if err != nil {
+		return nil, nil, err
+	}
+	var relevantBoards []*workitem.Board
+	for _, board := range boards {
+		// this rule is only dealing with TypeLevelContext boards right now
+		// this may need to be extended when we allow other boards.
+		if board.ContextType == "TypeLevelContext" {
```

Comment:
Maybe continue if this condition doesn't match? It would avoid the heavy nesting.

Reply:
This is reflecting my thinking process.

Reply:
@michaelkleinhenz @kwk This would be easier to read with less nesting. It makes it easier to read and much easier to maintain in the long run.